src/libraries/Common/src/SourceGenerators/DiagnosticInfo.cs

```diff
@@ -30,8 +35,35 @@ public static DiagnosticInfo Create(DiagnosticDescriptor descriptor, Location? l
         };
 
         // Creates a copy of the Location instance that does not capture a reference to Compilation.
-        static Location GetTrimmedLocation(Location location)
-            => Location.Create(location.SourceTree?.FilePath ?? "", location.SourceSpan, location.GetLineSpan().Span);
+        static Location? GetTrimmedLocation(Location? sourceLocation)
+        {
+            if (sourceLocation is null)
+            {
+                return null;
+            }
+
+            Location trimmedLocation = Location.Create(sourceLocation.SourceTree?.FilePath ?? "", sourceLocation.SourceSpan, sourceLocation.GetLineSpan().Span);
+
+            if (sourceLocation.IsInSource &&
+                !trimmedLocation.IsInSource &&
+                s_sourceTreeBackingFieldInfo.Value is FieldInfo sourceTreeField &&
+                s_dummySyntaxTree.Value is CSharpSyntaxTree syntaxTree)
```

Comment:
If the instance is only used to check if the `DummySyntaxTree` type exists, why not just store a `Type?` instance?